code/net/ConnectX-5/mlx5.c

```diff
@@ -0,0 +1,412 @@
+  // Thus we can ask for 128KB, since we don't know the flash size (and don't need it thus no need to actually check it)
+  if (!tn_mem_phys_to_virt(dev_phys_addr, 128 * 1024, &device.addr)) {
+    return false;
+  }
+
+  TN_VERBOSE("Device %02" PRIx8 ":%02" PRIx8 ".%" PRIx8 " mapped to %p", pci_address.bus, pci_address.device, pci_address.function, device.addr);
+
+  /**
+   * Step 1.
+   * Read the initialization segment from offset 0 of the HCA BAR,
+   * to retrieve the versions of the firmware and the command interface.
+   * The driver must match the command interface revision number.
+   * The format of the initialization segment is in Table 12, “Initialization Segment,” on page 168
+   * */
+  TN_VERBOSE("### Init - step 1");
+
+  uint32_t fw_rev = reg_read(device.addr, 0x00);
+  uint32_t cmd_rev_fw_rev_subminor = reg_read(device.addr, 0x04);
+  uint16_t fw_rev_minor = fw_rev >> 16;
+  uint16_t fw_rev_major = (uint16_t) fw_rev & 0x0000FFFF;
+  uint16_t cmd_int_rev = cmd_rev_fw_rev_subminor >> 16;
+  uint16_t fw_rev_subminor = (uint16_t) cmd_rev_fw_rev_subminor & 0x0000FFFF;
+
+  if (fw_rev_minor != FW_REV_MINOR) {
+    TN_DEBUG("Firmware revision minor is %x, expected value is %x", fw_rev_minor, FW_REV_MINOR);
+    return false;
+  }
+
+  if (fw_rev_major != FW_REV_MAJOR) {
+    TN_DEBUG("Firmware revision is major %x expected value is %x", fw_rev_major, FW_REV_MAJOR);
+    return false;
+  }
+
+  if (cmd_int_rev != CMD_INTERFACE_REVISION) {
+    TN_DEBUG("Firmware revision is %x expected value is %x", cmd_int_rev, CMD_INTERFACE_REVISION);
+    return false;
+  }
+
+  if (fw_rev_subminor != FW_REV_SUBMINOR) {
+    TN_DEBUG("Firmware revision is %x expected value is %x", fw_rev_subminor, FW_REV_SUBMINOR);
+    return false;
+  }
+
+  /**
+   * Step 2.
+   *  Write the physical location of the command queues to the initialization segment.
+   *  If using 32-bit writes, write the most significant word first.
+   *  The nic_interface field is part of the least significant word
+   *  and must be set to zero (Full NIC/HCA driver), as are the log_cmdq_size and log_cmdq_stride fields
+   * */
+  TN_VERBOSE("### Init - step 2");
+
+  void* command_queues_virt_addr;
+  uintptr_t command_queues_phys_addr;
+  TN_DEBUG("size of command_queues_phys_addr %ld", sizeof(command_queues_phys_addr));
+
+  TN_VERBOSE("Trying to allocate %d bytes for command queues", CMDQ_SIZE);
+  if (!tn_mem_allocate(CMDQ_SIZE, &command_queues_virt_addr)) {
+    TN_DEBUG("Could not allocate memory for command queues");
+    return false;
+  }
+
+  if (!tn_mem_virt_to_phys(command_queues_virt_addr, &command_queues_phys_addr)) {
+    TN_DEBUG("Could not get a command_queues's physical address");
+    return false;
+  }
+
+  reg_write(device.addr, 0x10, (uint32_t)command_queues_phys_addr>>32);
+  reg_write_field(device.addr, 0x14, CMDQ_PHY_ADDR_LOW, (uint32_t) command_queues_phys_addr & 0x00000000FFFFFFFF);
```

Comment:
Since `CMDQ_PHY_ADDR_LOW` is not 32 bits but 20, I think the lower 12 bits are implicitly 0, so you'll have to (1) double-check that the physical address indeed has the lower 12 bits set to 0 and (2) shift the address here to write only the top 20 bits of the bottom 32 bits (if that makes sense).

Reply:
Do I still need to shift with my current version of the code ?

Reply:
Yes, otherwise what'll happen is your 32-bit value will be cut off at the top so you'll write the lower bits of the address instead of the upper bits